src/ape/pytest/contextmanagers.py

```diff
@@ -1,26 +1,141 @@
+            exception.txn.txn_hash,
+            raise_on_fail=False,
+        )
+
+        try:
+            contract = ape.Contract(receipt.receiver)
+        except ValueError as exc:
             raise AssertionError(
-                f"Transaction did not revert.\n"
-                f"However, an exception of type {type(exc_value)} occurred: {exc_value}"
-            ) from exc_value
+                f"Could not fetch contract at {receipt.receiver} to check dev message."
+            ) from exc
+
+        if contract.contract_type.pcmap is None:
+            raise AssertionError("Compiler does not support source code mapping.")
+
+        try:
+            trace = deque(exception.trace)
+        except APINotImplementedError as exc:
+            raise AssertionError(
+                "Cannot check dev message; provider must support transaction tracing."
+            )
+
+        if trace[-1].op != "REVERT":
+            raise AssertionError("Transaction trace did not end with REVERT.")
+
+        pc = None
+        pcmap = contract.contract_type.pcmap.parse()
+
+        # To find a suitable line for inspecting dev messages, we must start at the revert and work
+        # our way backwards. If the last frame's PC is in the PC map, the offending line is very
+        # likely a 'raise' statement.
+        if trace[-1].pc in contract.contract_type.pcmap:
+            pc = trace[-1].pc
+
+        # Otherwise we must traverse the trace backwards until we find our first suitable candidate.
+        # It should be noted that when an error happens, typically the contract will execute:
+        #
+        #   <...>
+        #   JUMPI    <- jump to revert routine on a failing condition
+        #
+        #   <...>
+        #   JUMPDEST <- beginning of revert routine
+        #   PUSH1
+        #   DUP1
+        #   REVERT
+        #
+        # So we must step backwards until we find the initial instructions that caused the failing
+        # condition. This will not be the JUMPI or even an arithmetic instruction like DIV, but
+        # rather the starting instruction for a given line.
+        else:
+            while len(trace) > 0:
+                frame = trace.pop()
+                if frame.pc in pcmap:
+                    pc = frame.pc
+                    break
+
+        # We were unable to find a suitable PC that matched the compiler's map.
+        if pc is None:
+            raise AssertionError("Could not find line that caused revert.")
+
+        offending_source = pcmap[pc]
+
+        # The compiler PC map had PC information, but not source information.
+        if offending_source is None or offending_source.line_start is None:
+            raise AssertionError("Could not find line that caused revert.")
+
+        try:
+            contract_source_lines = (
+                ape.project.extract_manifest()
+                .sources[contract.contract_type.source_id]
+                .content.split("\n")
+            )
```

Comment:
Mypy questions:

```
Value of type "Optional[Dict[str, Source]]" is not indexable

Item "None" of "Union[str, None, Any]" has no attribute "split"`

Invalid index type "Optional[str]" for "Optional[Dict[str, Source]]"; expected type "str"
```

Do I need to be saving `sources` to like a typed variable? From what I can tell `sources` should always be `Dict[str, Source]`. Is there a situation where it's `None` or has null values?

cc @NotPeopling2day @unparalleled-js

Reply:
On the index one: when is `source_id` null? Would I run into that in this scenario somehow after already constructing the `Contract` earlier up in this function?

Reply:
`source_id` can be `None` if the source is not a local source, like you just got it from somewhere else and it wasn't included.

This happens sometimes when using hardcoded ABIs especially because there is no reason to use a source ID (until there is).

Reply:
`sources` is `None` is not using any local sources, as in you are only using dependencies.
The spec supports this, and so do we in Ape, that is why the type of `sources` is optional.

Reply:
@unparalleled-js Gotcha, thanks! Just updated to 53dce52c

I opted for just checking them first rather than the long statement in a try/except that I had:

<img width="713" alt="image" src="https://user-images.githubusercontent.com/108302637/199355307-3b12e162-4c83-4984-b754-47fd9c4594ed.png">

I noted in the PR description as well that this currently will only work with local sources, I think I could take a stab at non-local sources (if that's something we should be supporting) in a second PR or a second ticket